src/utility/HCISharedMemTransport.cpp

```diff
@@ -0,0 +1,837 @@
+  _write_index_initial = 0;
+
+  sys_event = false;
+  sys_resp = false;
+  acl_data = false;
+}
+
+static void evt_received(TL_EvtPacket_t *hcievt)
+{
+    uint16_t len = 0;
+
+    // We need to memcpy the data before passing to higher layers.
+    switch (hcievt->evtserial.type) {
+        case TL_BLEEVT_PKT_TYPE:
+            len = hcievt->evtserial.evt.plen + TL_EVT_HDR_SIZE;
+            //TODO: on_data_received((uint8_t *)&hcievt->evtserial, len);
+            break;
+        case TL_ACL_DATA_PKT_TYPE: {
+            TL_AclDataSerial_t *acl = &(((TL_AclDataPacket_t *)hcievt)->AclDataSerial);
+            len = acl->length + 5;
+            //TODO: on_data_received((uint8_t *)acl, len);
+        }
+        break;
+        default:
+            // should not happen - let's block to check
+#if defined(DEBUG)
+            tr_error("BLE TL evt_received, wrong type:%d", hcievt->evtserial.type);
+#endif /* DEBUG */
+            break;
+    }
+
+    /*  In case Event belongs to the Evt Pool we need to inform  */
+    if (((uint8_t *)hcievt >= EvtPool) && ((uint8_t *)hcievt < (EvtPool + POOL_SIZE))) {
+        /*  Free the message from shared memory */
+        TL_MM_EvtDone(hcievt);
+    }
+}
+
+/**
+ * TL Mailbox synchronisation means
+ */
+
+/*  returns true if sys_event was received, false otherwise */
+static bool sysevt_wait(void)
+{
+    /* sys_event remains false until event is received */
+//    for (unsigned long start = millis(); (millis() - start) < 10000;) {
+    while(1) {
+        /* Wait for 10sec max - if not return an error */
+        if (sys_event) { break; }
+    }
+
+    if (!sys_event) {
+        /*  no event received, timeout occurs */
+         return false;
+    } else {
+        /*  release immmediately, now that M0 runs */
+        sys_event = false;
+        return true;
+    }
+}
+
+/*  WEAK callbacks from the BLE TL driver - will be called under Interrupt */
+static void sysevt_received(void *pdata)
+{
+    /* For now only READY event is received, so we know this is it */
+    sys_event = true;
+    /* But later on ... we'll have to parse the answer */
+    
+    return;
+}
+
+/*  returns true if sysevt was already received, which means M0 core is
+ *  already up and running */
+static bool sysevt_check(void)
+{
+    /*  Check if system is UP and runing already */
+//    for (unsigned long start = millis(); (millis() - start) < 10;) {
+    while(1) {
+        if (sys_event) { break; }
+    }
+
+    if (!sys_event) {
+        return false;
+    } else {
+        /*  release immmediately as M0 already runs */
+        sys_event = false;
+        return true;
+    }
+}
+
+void shci_cmd_resp_release(uint32_t flag)
+{
+    sys_resp = false;
+    return;
+}
+
+void shci_cmd_resp_wait(uint32_t timeout)
+{
+    /* TO DO: manage timeouts if we can return an error */
+//    for (unsigned long start = millis(); (millis() - start) < 10000;) {
+    while(1) {
+        /* Wait for 10sec max - if not return an error */
```

Comment:
Here if the variable is already false, wait for it to become true. 
Once true (or if already true), set to false and return ok.